source/common/buffer/buffer_impl.h

```diff
@@ -749,6 +753,30 @@ class OwnedImpl : public LibEventInstance {
   void commit(uint64_t length, absl::Span<RawSlice> slices,
               ReservationSlicesOwnerPtr slices_owner) override;
 
+  uint8_t* inlineReserve(uint64_t size) override {
+    // If no any slice in the buffer then create one new slice based on the expected continuous
+    // memory size.
+    if (slices_.empty()) {
+      slices_.emplace_back(Slice(size, account_));
+      length_ += size;
+      slices_.back().reservable_ += size;
+      return slices_.back().base_;
+    }
+
+    // If the back slice has enough continuous memory then reserve the memory segment and mark it
+    // as already used.
+    auto& back = slices_.back();
+    if (back.reservableSize() >= size) {
+      uint8_t* mem = back.base_ + back.reservable_;
+      length_ += size;
+      back.reservable_ += size;
+      return mem;
+    }
+
+    // No enough continuous memory in the back slice.
+    return nullptr;
```

Comment:
> Why not allocate a slice like in the empty case?

Because the method is designed to reserve continuous memory from exist back slice. And there are detailed comments in the method descriptions.
```
  /**
   * Reserve contiguous memory segment with specific size from the back slice. The first address of
   * this memory segment will be returned. The reserved memory will be regarded as used and directly
   * increase the length of the buffer.
   *
   * @param expected contiguous memory segment size.
   * @return the first address of this memory segment or nullptr if there is not enough continuous
   * memory in the back slice.
   */
```

And a step further:

> Why did we design the method the way it is?

There is much of the previous discussion and benchmark results in the PR. **The key issue is avoiding invalid memory allocations and the performance degradation caused by invalid memory allocations**.

Here are some previous comments:

Why create new slice when back slice has no enough continuous memory is not good choice: https://github.com/envoyproxy/envoy/pull/19115#issuecomment-995402699

Why new design like this: https://github.com/envoyproxy/envoy/pull/19115#issuecomment-1001510577

Reply:
What do you mean by "invalid memory allocations"?

Please comment/document in the code the rationale for why it is this way. If someone reading it doesn't understand why, they may change it in a way that simplifies the code and reverts the performance improvements you're making.

Reply:
For example, if we want write five 1024 bytes strings to buffer by single `addFragments` call. We may reserve 5kb continuous memory for no-intermediate-check writes. Then we will create a new 8kb slice for this write.
But notice that the next time `addFragments` may writes five 1024 bytes strings again, the buffer may need creates a new 8kb slice again because the previous slice does not have enough contiguous memory. In this case, 3kb of space in the first slice is wasted.


But we has figured out a new simpler and cleaner implementation (Thanks for cc @jmarantz @antoniovicente 's suggestions). So this code is outdated. You can check the latest commit for the latest implementation. 😄